Clarify canonical format used for Breadcrumbs, Exceptions and Threads (#1554)
getsentry/sentry-docs: src/collections/_documentation/development/sdk-dev/event-payloads/breadcrumbs.md

sidebar_order: 6
---

The Breadcrumbs Interface specifies a series of application events, or

> "breadcrumbs", that occurred before and led up to an event.

Reviewer: ```suggestion
"breadcrumbs", that occurred before an event.
```

An [event]({%- link _documentation/development/sdk-dev/event-payloads/index.md
-%}) may contain one or more breadcrumbs in an attribute named `breadcrumbs`.

This interface is an object with a single `values` attribute containing an

> ordered list of breadcrumb objects. The entries are ordered from oldest to

Reviewer: ```suggestion
ordered list of breadcrumb objects. The list will order the entries from oldest to newest. 
```

Author: @MimiDumpling please help me with this one.

In "The list will order the entries from oldest to newest", I understand that "The list" is the subject, "will order" the verb, "the entries" the object.

Given that a list is a data structure that contains an ordered sequence of items, "the list" itself doesn't "order"(/"sort"?) anything. The items are intrinsically in some order.

Should I interpret the sentence differently?

As similar sentence appears in `exceptions.md`, https://github.com/getsentry/sentry-docs/pull/1554/files#diff-9b11e393b13bda7b8428e2bdd3d87becR15-R16. We should align those.

Reviewer: @rhcarvalho Yeah, this was my attempt at changing the sentence to active voice by giving the sentence a clear subject. If it doesn't make technical sense, we can disregard my changes here. 😃

> newest. Consequently, the last entry in the list should be the last entry before

Reviewer: ```suggestion
 Consequently, the last entry in the list should be the last entry before
```

Reviewer: @rhcarvalho I wasn't trying to force a line break. Wanted to delete "newest" because I added it to the previous line.

Author: Ah, as far as rendering goes, those two render the same:

```md
A long line with a line
break. Another sentence.
```

```md
A long line with a line break.
Another sentence.
```

The result is a single `<p>` HTML tag displaying all in one line/paragraph:

```
A long line with a line break. Another sentence.
```

When editing markdown it is customary to wrap long lines. To start a new paragraph we need a blank line between blocks of text, and to insert a line break (`<br>` HTML tag) we need to end a line with two (!) (or more) spaces.

https://daringfireball.net/projects/markdown/syntax#p

Sorry if I'm preaching to the priest, probably you know all that. 😊

Reviewer: All good. 👍

the event occurred.

> Each breadcrumb has a few properties of which at least `timestamp` and

Reviewer: ```suggestion
Each breadcrumb has a few properties you must provide, such as `timestamp` and
```

Reviewer: @rhcarvalho Yeah, on line 18 I deleted "must be provided." Does it not show up?

Reviewer: It should read "Each breadcrumb has a few properties you mush provide such as `timestamp` and `category`."

Author: > Does it not show up?

I missed it right under my nose 🤦‍♂ 

I'll have to check the implementation because here we say `timestamp` and `category` are required, but then in the [list of attributes](https://docs.sentry.io/development/sdk-dev/event-payloads/breadcrumbs/#attributes) we say `category` is `optional`.

Author: @jan-auer or @untitaker could you please help clarify what is actually required for a _breadcrumb_?

Part of the docs says `timestamp` and `category`, part says only `timestamp`. In practice, nothing seems to be strictly required.

I looked at the code and sent a sample event. In practice, it works to send an empty breadcrumb `{"breadcrumbs": {"values": [{}]}`, which the server [normalizes](https://github.com/getsentry/relay/blob/5069fbc7846782f372bfc0dee522e0aae7d9b673/relay-general/src/store/normalize.rs#L474-L491) as `{"breadcrumbs":{"values":[{"type":"default","level":"info"}]}}`.

Author: This is how an "empty" breadcrumb like described in my previous comment renders:

![image](https://user-images.githubusercontent.com/88819/78020204-4b3c2700-7351-11ea-8d18-968a76835a26.png)

The breadcrumb is in the "generic" line.

Author: After talking with Markus, decided to update the docs to reflect the truth, i.e., nothing is strictly required. We can change the wording from **required** to **recommended**, as otherwise the value from the breadcrumb is rather limited, and in some cases the non-timestamped information could be sent in a different field (e.g. a tag).

Author: Since this was not part of the original diff, I'll make changes to what is required/recommended in a new PR.

Author: Sent https://github.com/getsentry/sentry-docs/pull/1579 to clarify what is required/recommended/optional.

> `category` must be provided. The rest is optional, and depending on what is

Reviewer: ```suggestion
`category`. The rest is optional, and depending on what you've
```

sidebar_order: 6
---

The Breadcrumbs Interface specifies a series of application events, or
"breadcrumbs", that occurred before and led up to an event.

An [event]({%- link _documentation/development/sdk-dev/event-payloads/index.md
-%}) may contain one or more breadcrumbs in an attribute named `breadcrumbs`.

This interface is an object with a single `values` attribute containing an
ordered list of breadcrumb objects. The entries are ordered from oldest to
newest. Consequently, the last entry in the list should be the last entry before
the event occurred.

{% capture __alert_content -%}
For backwards compatibility, the server also understands events with exceptions

> that are not in the canonical format described here.

Reviewer 2: This talks about exceptions - copy/paste error?

Author: And removed preferring a more global note, as per Jan's suggestion.

## Examples

The following example illustrates the breadcrumbs part of the [event
payload]({%- link _documentation/development/sdk-dev/event-payloads/index.md

> -%}) and omit other attributes for simplicity.

Reviewer: ```suggestion
-%}) and omits other attributes for simplicity.
```
